src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -874,17 +877,98 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
                     auto copy = cmdImpl->Copy();
                     copy->NestedCommands(innerResults.GetView());
 
-                    results.Append(*copy);
+                    results.push_back(*copy);
                 }
             }
+        };
+
+        // iterate over all the commands in all our actions...
+        for (auto&& [name, command] : nameMap)
+        {
+            addCommand(command);
+        }
+        // ... and all the local commands passed in here
+        for (const auto& command : localCommands)
+        {
+            addCommand(command);
         }
 
         return results;
     }
 
-    IVector<Model::Command> ActionMap::FilterToSendInput(
-        winrt::hstring currentCommandline)
+    // Update ActionMap's cache of actions for this directory. We'll look for a
+    // .wt.json in this directory. If it exists, we'll read it, parse it's JSON,
+    // then take all the sendInput actions in it and store them in our
+    // _cwdLocalSnippetsCache
+    winrt::Windows::Foundation::IAsyncAction ActionMap::_updateLocalSnippetCache(winrt::hstring currentWorkingDirectory)
     {
-        return _filterToSendInput(NameMap(), currentCommandline);
+        // Don't do I/O on the main thread
+        co_await winrt::resume_background();
+
+        // This returns an empty string if we fail to load the file.
+        auto localTasksFileContents = CascadiaSettings::ReadFile(currentWorkingDirectory + L"\\.wt.json");
+        if (!localTasksFileContents.empty())
+        {
+            auto data = winrt::to_string(localTasksFileContents);
+            std::string errs;
+            static std::unique_ptr<Json::CharReader> reader{ Json::CharReaderBuilder::CharReaderBuilder().newCharReader() };
+            Json::Value root;
+            if (!reader->parse(data.data(), data.data() + data.size(), &root, &errs))
+            {
+                // In the real settings parser, we'd throw here:
+                // throw winrt::hresult_error(WEB_E_INVALID_JSON_STRING, winrt::to_hstring(errs));
+                //
+                // That seems overly aggressive for something that we don't
+                // really own. Instead, just bail out.
+                co_return;
+            }
+
+            auto result = std::vector<Model::Command>();
+            if (auto actions{ root[JsonKey("actions")] })
+            {
+                std::vector<SettingsLoadWarnings> warnings;
+                for (const auto& json : actions)
+                {
+                    auto parsed = Command::FromJson(json, warnings, OriginTag::Generated);
+                    // Skip over things that aren't snippets
+                    if (parsed->ActionAndArgs().Action() != ShortcutAction::SendInput)
+                    {
+                        continue;
+                    }
+
+                    result.push_back(*parsed);
+                }
+            }
+
+            _cwdLocalSnippetsCache.insert_or_assign(currentWorkingDirectory, result);

```

Comment:
Yes, this isn't thread-safe. Even if this function doesn't get called concurrently, without a mutex or similar there's no guarantee that the UI thread sees this data in a consistent state - it may read stale data, which may already be deallocated for instance.

You could use a `mutex` here, but as I mentioned before, heap allocations are essentially the exact same thing as reading a file from disk: system memory may be backed by the swap, aka `pagefile.sys`, after all. So, in a way, `resume_foreground` is the "right way". However, considering how WinUI guzzles heap memory like a Ford F150 rolling coal, I think that is the least of our worries. mutexes are probably fine!

Reply:
Done, as discussed in chat